File: 0_字符串/字符串变换/917_仅仅反转字母.ts
import { isalpha } from '../../utils/string'

function reverseOnlyLetters(s: string): string {
  const sb = s.split('')
  let [left, right] = [0, s.length - 1]

  while (left < right) {
    while (left < right && !isalpha(sb[left])) left++
    while (left < right && !isalpha(sb[right])) right--
    if (left < right) {
      ;[sb[left], sb[right]] = [sb[right], sb[left]]
      left++
      right--
    }
  }

  return sb.join('')
}

File: 0_字符串/字符串变换/917_仅仅反转字母.py
class Solution:
    def reverseOnlyLetters(self, s: str) -> str:
        sb = list(s)
        left, right = 0, len(s) - 1
        while left < right:
            while left < right and not s[left].isalpha():
                left += 1
            while left < right and not s[right].isalpha():
                right -= 1
            if left < right:
                sb[left], sb[right] = s[right], s[left]
                left += 1
                right -= 1

        return "".join(sb)


print(Solution().reverseOnlyLetters("a-bC-dEf-ghIj"))
